generate/templates/manual/revwalk/file_history_walk.cc

```diff
@@ -34,251 +214,222 @@ NAN_METHOD(GitRevwalk::FileHistoryWalk)
         }
-      }
-
-      flag = false;
-      numDeltas = git_diff_num_deltas(diffs);
-      for (unsigned int j = 0; j < numDeltas; ++j) {
-        git_patch *nextPatch;
-        baton->error_code = git_patch_from_diff(&nextPatch, diffs, j);
-
-        if (baton->error_code < GIT_OK) {
+        case GIT_DELTA_RENAMED: {
+          if (fileHistoryEvent->existsInCurrentTree) {
+            fileExistsInCurrent = true;
+          } else {
+            fileExistsInSomeParent = true;
+          }
           break;
         }
-
-        if (nextPatch == NULL) {
-          continue;
-        }
-
-        const git_diff_delta *delta = git_patch_get_delta(nextPatch);
-        bool isEqualOldFile = !strncmp(delta->old_file.path, baton->file_path, strlen(baton->file_path));
-        bool isEqualNewFile = !strncmp(delta->new_file.path, baton->file_path, strlen(baton->file_path));
-        int oldLen = strlen(delta->old_file.path);
-        int newLen = strlen(delta->new_file.path);
-        char *outPair = new char[oldLen + newLen + 2];
-        strcpy(outPair, delta->new_file.path);
-        outPair[newLen] = '\n';
-        outPair[newLen + 1] = '\0';
-        strcat(outPair, delta->old_file.path);
-
-        if (isEqualNewFile) {
-          std::pair<git_commit *, std::pair<char *, git_delta_t> > *historyEntry;
-          if (!isEqualOldFile || delta->status == GIT_DELTA_RENAMED) {
-            historyEntry = new std::pair<git_commit *, std::pair<char *, git_delta_t> >(
-              nextCommit,
-              std::pair<char *, git_delta_t>(strdup(outPair), delta->status)
-            );
-          } else {
-            historyEntry = new std::pair<git_commit *, std::pair<char *, git_delta_t> >(
-              nextCommit,
-              std::pair<char *, git_delta_t>(strdup(delta->new_file.path), delta->status)
-            );
+        case GIT_DELTA_UNMODIFIED: {
+          if (fileHistoryEvent->existsInCurrentTree) {
+            fileExistsInCurrent = true;
+            fileExistsInSomeParent = true;
           }
-          baton->out->push_back(historyEntry);
-          flag = true;
-        } else if (isEqualOldFile) {
-          std::pair<git_commit *, std::pair<char *, git_delta_t> > *historyEntry;
-          historyEntry = new std::pair<git_commit *, std::pair<char *, git_delta_t> >(
-            nextCommit,
-            std::pair<char *, git_delta_t>(strdup(outPair), delta->status)
-          );
-          baton->out->push_back(historyEntry);
-          flag = true;
+          firstMatchingParentIndex = std::make_pair(true, parentIndex);
+          break;
         }
-
-        delete[] outPair;
-
-        git_patch_free(nextPatch);
-
-        if (flag) {
+        default: {
           break;
         }
       }
-    }
 
-    git_diff_free(diffs);
+      delete fileHistoryEvent;
 
-    if (!flag && nextCommit != NULL) {
-      git_commit_free(nextCommit);
+      if (firstMatchingParentIndex.first) {
+        git_commit_free(parentCommit);
```

Comment:
These need to be freed at the same time we delete fileHistoryEvent or else we're leaking memory on non-matching on merge commits